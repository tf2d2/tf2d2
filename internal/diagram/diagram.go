/*
Copyright Â© 2023 The tf2d2 Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package diagram

import (
	"context"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"strings"

	iconsAWS "github.com/tf2d2/icons/providers/aws"
	iconsAzure "github.com/tf2d2/icons/providers/azurerm"
	iconsGoogle "github.com/tf2d2/icons/providers/google"
	d2tpl "github.com/tf2d2/tf2d2/internal/template"

	"github.com/cycloidio/inframap/graph"
	"github.com/hashicorp/go-hclog"

	"oss.terrastruct.com/d2/d2format"
	"oss.terrastruct.com/d2/d2graph"
	"oss.terrastruct.com/d2/d2layouts/d2dagrelayout"
	"oss.terrastruct.com/d2/d2lib"
	"oss.terrastruct.com/d2/d2renderers/d2svg"
	"oss.terrastruct.com/d2/d2target"
	"oss.terrastruct.com/d2/d2themes/d2themescatalog"
	"oss.terrastruct.com/d2/lib/png"
	"oss.terrastruct.com/d2/lib/textmeasure"
	"oss.terrastruct.com/util-go/go2"
	"oss.terrastruct.com/util-go/xmain"
)

// IDiagram defines how to generate a d2 diagram
type IDiagram interface {
	Initialize() error
	Generate() error
	Render() error
}

// Diagram implements IDiagram to generate a d2 diagram
type Diagram struct {
	ctx           context.Context // parent context
	Filepath      string
	TFInfraMap    *graph.Graph
	d2Graph       *d2graph.Graph
	d2CompileOpts *d2lib.CompileOptions
	d2RenderOpts  *d2svg.RenderOpts
}

// NewDiagram creates a new Diagram instance
func NewDiagram(ctx context.Context, m *graph.Graph, filepath string) *Diagram {
	return &Diagram{
		ctx:           ctx,
		Filepath:      filepath,
		TFInfraMap:    m,
		d2Graph:       nil,
		d2CompileOpts: nil,
		d2RenderOpts:  nil,
	}
}

// Initialize creates a new empty D2 graph
func (d *Diagram) Initialize() error {
	logger := hclog.FromContext(d.ctx)

	// initialize a ruler to measure font glyphs
	ruler, err := textmeasure.NewRuler()
	if err != nil {
		return err
	}

	// initialize layout resolver
	layoutResolver := func(engine string) (d2graph.LayoutGraph, error) {
		return d2dagrelayout.DefaultLayout, nil
	}

	// initialize compile options
	compileOpts := &d2lib.CompileOptions{
		LayoutResolver: layoutResolver,
		Ruler:          ruler,
	}

	// initialize render options
	renderOpts := &d2svg.RenderOpts{
		Pad:     go2.Pointer[int64](d2svg.DEFAULT_PADDING),
		Sketch:  go2.Pointer[bool](false),
		Center:  go2.Pointer[bool](false),
		ThemeID: go2.Pointer[int64](d2themescatalog.NeutralDefault.ID),
	}

	// initialize empty diagram
	_, graph, err := d2lib.Compile(d.ctx, "", nil, nil)
	if err != nil {
		return fmt.Errorf("error initializing diagram: %w", err)
	}

	d.d2Graph = graph
	d.d2CompileOpts = compileOpts
	d.d2RenderOpts = renderOpts

	logger.Debug("initialized d2 diagram")

	return nil
}

// Generate generates a D2 diagram
func (d *Diagram) Generate(dryRun bool) error {
	logger := hclog.FromContext(d.ctx)

	// compute d2 shapes
	shapes := []*d2target.Shape{}
	for _, n := range d.TFInfraMap.Nodes {
		logger.Debug(fmt.Sprintf("%#v\n", n))

		s := d2target.BaseShape()
		s.ID = strings.ReplaceAll(n.Canonical, ".", "_")
		s.Label = n.Resource.Name

		iconURL, err := getIconURL(n.Resource.Name)
		if err != nil {
			return err
		}
		s.Icon = iconURL
		shapes = append(shapes, s)
	}

	// compute d2 connections between shapes
	conns := []*d2target.Connection{}
	for _, e := range d.TFInfraMap.Edges {
		sourceN, err := d.TFInfraMap.GetNodeByID(e.Source)
		if err != nil {
			return fmt.Errorf("error getting source node: %w", err)
		}
		targetN, err := d.TFInfraMap.GetNodeByID(e.Target)
		if err != nil {
			return fmt.Errorf("error getting target node: %w", err)
		}

		c := d2target.BaseConnection()
		c.Src = strings.ReplaceAll(sourceN.Canonical, ".", "_")
		c.Dst = strings.ReplaceAll(targetN.Canonical, ".", "_")
		c.DstArrow = d2target.ArrowArrowhead
		conns = append(conns, c)
	}

	// render d2 template with computed shapes and connections
	t := d2tpl.New(shapes, conns)
	out, err := t.Render("d2")
	if err != nil {
		return err
	}

	// compile d2 diagram from rendered template output
	_, d.d2Graph, err = d2lib.Compile(d.ctx, out, d.d2CompileOpts, d.d2RenderOpts)
	if err != nil {
		return err
	}

	// render d2 diagram
	err = d.Write(dryRun)
	if err != nil {
		return err
	}

	return nil
}

// Write creates a file for both the rendered d2 diagram and compiled d2 script
// provided it's not a dry run. Otherwise, the d2 script is written to stdout
func (d *Diagram) Write(dryRun bool) error {
	// turn the graph into a script
	script := d2format.Format(d.d2Graph.AST)

	if dryRun {
		_, err := os.Stdout.WriteString(script + "\n")
		if err != nil {
			return fmt.Errorf("error writing to standard output: %w", err)
		}
	} else {
		// compile the script into a diagram
		diagram, _, err := d2lib.Compile(d.ctx, script, d.d2CompileOpts, d.d2RenderOpts)
		if err != nil {
			return err
		}

		// render to svg
		var svgData []byte
		svgData, err = d2svg.Render(diagram, d.d2RenderOpts)
		if err != nil {
			return fmt.Errorf("error rendering svg data: %w", err)
		}

		// write output files
		err = writeContent(d.Filepath, script, svgData)
		if err != nil {
			return fmt.Errorf("error writing output to disk: %w", err)
		}
	}

	return nil
}

func writeContent(path string, scriptData string, svgData []byte) error {
	fileExtension := filepath.Ext(path)
	switch fileExtension {
	case ".png":
		// write PNG diagram to output file path
		// chromium is used to convert SVG to PNG, it's installed if not available
		var pw png.Playwright
		pw, err := png.InitPlaywright()
		if err != nil {
			return err
		}
		defer func() {
			cleanUpErr := pw.Cleanup()
			if err != nil {
				err = cleanUpErr
			}
		}()

		var pngData []byte
		pngData, err = png.ConvertSVG(&xmain.State{}, pw.Page, svgData)
		if err != nil {
			return err
		}

		err = os.WriteFile(path, pngData, 0600)
		if err != nil {
			return err
		}
	default:
		// write SVG diagram to output file path
		err := os.WriteFile(path, svgData, 0600)
		if err != nil {
			return err
		}
	}

	// write d2 script to output file path
	scriptFilepath := strings.ReplaceAll(path, fileExtension, ".d2")
	err := os.WriteFile(scriptFilepath, []byte(scriptData), 0600)
	if err != nil {
		return err
	}

	return nil
}

func getIconURL(resource string) (*url.URL, error) {
	var iconURL *url.URL
	prefix, _, _ := strings.Cut(resource, "_")
	switch prefix {
	case "aws":
		r, err := iconsAWS.GetResource(resource)
		if err != nil {
			return nil, err
		}
		iconURL, err = url.Parse(r.IconURL)
		if err != nil {
			return nil, err
		}
	case "azurerm":
		r, err := iconsAzure.GetResource(resource)
		if err != nil {
			return nil, err
		}
		iconURL, err = url.Parse(r.IconURL)
		if err != nil {
			return nil, err
		}
	case "google":
		r, err := iconsGoogle.GetResource(resource)
		if err != nil {
			return nil, err
		}
		iconURL, err = url.Parse(r.IconURL)
		if err != nil {
			return nil, err
		}
	default:
		return &url.URL{}, nil
	}

	return iconURL, nil
}
