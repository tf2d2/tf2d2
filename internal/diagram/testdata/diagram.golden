<?xml version="1.0" encoding="utf-8"?><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" d2Version="v0.6.0-HEAD" preserveAspectRatio="xMinYMin meet" viewBox="0 0 550 521"><svg id="d2-svg" class="d2-827229399" width="550" height="521" viewBox="-101 -101 550 521"><rect x="-101.000000" y="-101.000000" width="550.000000" height="521.000000" rx="0.000000" class=" fill-N7" stroke-width="0" /><style type="text/css"><![CDATA[
.d2-827229399 .text-bold {
	font-family: "d2-827229399-font-bold";
}
@font-face {
	font-family: d2-827229399-font-bold;
	src: url("data:application/font-woff;base64,d09GRgABAAAAAApYAAoAAAAAEIQAAguFAAAAAAAAAAAAAAAAAAAAAAAAAABPUy8yAAAA9AAAAGAAAABgXxHXrmNtYXAAAAFUAAAAbwAAAI4CRAMXZ2x5ZgAAAcQAAARaAAAFrJK8eSJoZWFkAAAGIAAAADYAAAA2G38e1GhoZWEAAAZYAAAAJAAAACQKfwXSaG10eAAABnwAAABMAAAATCPmAudsb2NhAAAGyAAAACgAAAAoDm4P7G1heHAAAAbwAAAAIAAAACAAKwD3bmFtZQAABxAAAAMoAAAIKgjwVkFwb3N0AAAKOAAAAB0AAAAg/9EAMgADAioCvAAFAAACigJYAAAASwKKAlgAAAFeADIBKQAAAgsHAwMEAwICBGAAAvcAAAADAAAAAAAAAABBREJPACAAIP//Au7/BgAAA9gBESAAAZ8AAAAAAfAClAAAACAAA3icZMzPqgEBAEfhb+7MvdefwSy8nyRFWUhKth6CSHgFT/hTo2yc5Vl8KJQK1CoHjDVKlYmpmbmlta2dfcLnLqxs3jfPPHLPLddccs4px9b7rlH4ae1ff/51dPX01QaGRrwAAAD//wEAAP//SHobTAB4nGRUTWzbVBz/vxfHb0m9ZU5iO0mbz1fbSbq4TVzby9rOzZa13dau3aq1G/so7ACDbh3aOlpNSLtMCBATgvSAOMAFJA7jMHGBSQWJwwaC2zbtAhIINHHMoeKUOchJtrbikDzLh9+3H3hhBgBfwGvgAR8EIAgCgM6neVlXVUos3bKo5LFUxJMZHHS+/ELNMbkck099kryxsICmzuO1Z5fOTF248O/C0JDz2Xf3nNvo2j0ADPnmBnqEGhAFCiBlFGPQtBSFZliimqZeEgWeqpRlrZJpGSwrhMXvqzO3apjmkqO9Rv/ivoVXV/1McnxHVA4dG05y8/axU4G0GhFeifcuXXX+1nvoVSk07++LRyQAQNDb3EDrqAExAG9GcelcFom4lEJY1EumJbEsih66Upl4q6qN9xyiKcO2ByJaaJ88x41cPzG7PJKQFuKTldEpIfByqhvA9eHi/okaEIHkNmRRCLMkLYp6ycX16IMuEUqOXz1w8NLQ+Ll+BjtP/GNFwywq5z/9Rt2TMbn9yyeOL9v2YjUk+0w9fTqWQPtyRj+09FdcMrwOYTd/XSDPQ+JbwISv1EjP0dLxw7V4qicbwet3Tkf7Fs85v6K0mY1Kzl1oNsECgN/xQ6wADwAEgvB+G7u5gYJ4HQLtJnidfxHIT5NDNd7nJWyQk7kzRzF99kQKInTZS9qaPHHUgHRLk6S3I92mjLw4K25fY0WjEkofKc4crcVT8oD714/qo8lCXzZTfC53wLnbOZ77Ro2O7w7HVt+rfiY19cI4qtuJwjbf7Y4wQQ0IQPf/OmLVkmkMdlaARPtKtXrFtpeq1SW7oGkFrVDodD+yPHvi+sjK1Ghl0p2Ai1tpTmARNSAECQBpU12YZWlGUSUh5GLTDBFE0dUZP6y+dHF4wUwNx7zTijnXlw9nv8VfFWP0vWsnV+3u6PRHqHds8p3CL8Fdnc7Rh6gBwW35EmXTefekIvT4Izuju3tGwqg+Xyp6vTcZJldy/gAEQnMDfY4aoLZ6VS13ja5ZRdWwMbgJJoRFKYGFMPuw+JpyIGMn04m4FksMZV8/WZ5PHogNxsplJTWSu8gpybPRbinEiyE/11vOHZpTI6fCohqJ7uqiZe3gufZW+eYGWsLLILXSNgxqWJYu6ALd8qHB2enqJH9jZYXGuahfClncG3M/X2Zv3br2IC+zzCLLtbEiALiO6q2NeXRJFN0YLGvLk4eqiuLeFYSsvf3xAOtnGbLTZ93c6wsQhvhI/7srdwpkJ2FIF9mD6k/lCUU5Qp+2zgn5qbP7Ph3LZsfo/RYf19yPnqG6u5TNzCxrK7VnF14V04EYCe6Qs37yw9p4V9DP7OB9w7fvSHunf2SZN5G3Nx5Dfz3OjMl0nD52uvafzEMHfxn903wAHgDJSAsc+u2D2dl21/AI1d33Oq/zlRqqO7sBNb/GZZjFD6ELgG/dku2ByZomy5qGy3lK8+4P/gMAAP//AQAA//+WUCYdAAAAAQAAAAILhfr2lWFfDzz1AAED6AAAAADYXaCEAAAAAN1mLzb+N/7ECG0D8QABAAMAAgAAAAAAAAABAAAD2P7vAAAImP43/jcIbQABAAAAAAAAAAAAAAAAAAAAEwKyAFACDwAqAdMAJAIGACQCOwBBARQANwEeAEEDWQBBAjwAQQIrACQCPQBBAY4AQQG7ABUBfwARAwgAGAIJAAwB9AAMARQAQQAA/60AAAAsAGQAkADEAOYA8gEOAUABYgGOAb4B3gIaAkACeAKoArQCwALWAAEAAAATAJAADABjAAcAAQAAAAAAAAAAAAAAAAAEAAN4nJyUz24bVRTGf05s0wrBAkVVuonugkWR6NhUSdU2K4fUikUUB48LQkJIE8/4jzKeGXkmDuEJWPMWvEVXPATPgVij+Xzs2AXRJoqSfHfu+fOdc75zgR3+ZptK9SHwRz0xXGGvfm54iwf1E8PbtOtbhqs8qf1puEZYmxuu83mtZ/gj3lZ/M/yA/epPhh+yW20b/phn1R3Dn2w7/jL8Kfu8XeAKvOBXwxV2yQxvscOPhrd5hMWsVHlE03CNz9gzXGcP6DOhIGZCwgjHkAkjrpgRkeMTMWPCkIgQR4cWMYW+JgRCjtF/fg3wKZgRKOKYAkeMT0xAztgi/iKvlHNlHOo0s7sWBWMCLuRxSUCCI2VESkLEpeIUFGS8okGDnIH4ZhTkeORMiPFImTGiQZc2p/QZMyHH0VakkplPypCCawLld2ZRdmZAREJurK5ICMXTiV8k7w6nOLpksl2PfLoR4Usc38m75JbK9is8/bo1Zpt5l2wC5upnrK7EurnWBMe6LfO2+Fa44BXuXv3ZZPL+HoX6XyjyBVeaf6hJJWKS4NwuLXwpyHePcRzp3MFXR76nQ58Turyhr3OLHj1anNGnw2v5dunh+JouZxzLoyO8uGtLMWf8gOMbOrIpY0fWn8XEIn4mM3Xn4jhTHVMy9bxk7qnWSBXefcLlDqUb6sjlM9AelZZO80u0ZwEjU0UmhlP1cqmN3PoXmiKmqqWc7e19uQ1z273lFt+QaodLtS44lZNbMHrfVL13NHOtH4+AkJQLWQxImdKg4Ea8zwm4IsZxrO6daEsKWiufMs+NVBIxFYMOieLMyPQ3MN34xn2woXtnb0ko/5Lp5aqq+2Rx6tXtjN6oe8s737ocrU2gYVNN19Q0ENfEtB9pp9b5+/LN9bqlPOWIlJjwXy/AMzya7HPAIWNlGOhmbq9DUy9Ek5ccqvpLIlkNpefIIhzg8ZwDDnjJ83f6uGTijItbcVnP3eKYI7ocflAVC/suR7xeffv/rL+LaVO1OJ6uTi/uPcUnd1DrF9qz2/eyp4mVk5hbtNutOCNgWnJxu+s1ucd4/wAAAP//AQAA///0t09ReJxiYGYAg//nGIwYsAAAAAAA//8BAAD//y8BAgMAAAA=");
}]]></style><style type="text/css"><![CDATA[.shape {
  shape-rendering: geometricPrecision;
  stroke-linejoin: round;
}
.connection {
  stroke-linecap: round;
  stroke-linejoin: round;
}
.blend {
  mix-blend-mode: multiply;
  opacity: 0.5;
}

		.d2-827229399 .fill-N1{fill:#0A0F25;}
		.d2-827229399 .fill-N2{fill:#676C7E;}
		.d2-827229399 .fill-N3{fill:#9499AB;}
		.d2-827229399 .fill-N4{fill:#CFD2DD;}
		.d2-827229399 .fill-N5{fill:#DEE1EB;}
		.d2-827229399 .fill-N6{fill:#EEF1F8;}
		.d2-827229399 .fill-N7{fill:#FFFFFF;}
		.d2-827229399 .fill-B1{fill:#0D32B2;}
		.d2-827229399 .fill-B2{fill:#0D32B2;}
		.d2-827229399 .fill-B3{fill:#E3E9FD;}
		.d2-827229399 .fill-B4{fill:#E3E9FD;}
		.d2-827229399 .fill-B5{fill:#EDF0FD;}
		.d2-827229399 .fill-B6{fill:#F7F8FE;}
		.d2-827229399 .fill-AA2{fill:#4A6FF3;}
		.d2-827229399 .fill-AA4{fill:#EDF0FD;}
		.d2-827229399 .fill-AA5{fill:#F7F8FE;}
		.d2-827229399 .fill-AB4{fill:#EDF0FD;}
		.d2-827229399 .fill-AB5{fill:#F7F8FE;}
		.d2-827229399 .stroke-N1{stroke:#0A0F25;}
		.d2-827229399 .stroke-N2{stroke:#676C7E;}
		.d2-827229399 .stroke-N3{stroke:#9499AB;}
		.d2-827229399 .stroke-N4{stroke:#CFD2DD;}
		.d2-827229399 .stroke-N5{stroke:#DEE1EB;}
		.d2-827229399 .stroke-N6{stroke:#EEF1F8;}
		.d2-827229399 .stroke-N7{stroke:#FFFFFF;}
		.d2-827229399 .stroke-B1{stroke:#0D32B2;}
		.d2-827229399 .stroke-B2{stroke:#0D32B2;}
		.d2-827229399 .stroke-B3{stroke:#E3E9FD;}
		.d2-827229399 .stroke-B4{stroke:#E3E9FD;}
		.d2-827229399 .stroke-B5{stroke:#EDF0FD;}
		.d2-827229399 .stroke-B6{stroke:#F7F8FE;}
		.d2-827229399 .stroke-AA2{stroke:#4A6FF3;}
		.d2-827229399 .stroke-AA4{stroke:#EDF0FD;}
		.d2-827229399 .stroke-AA5{stroke:#F7F8FE;}
		.d2-827229399 .stroke-AB4{stroke:#EDF0FD;}
		.d2-827229399 .stroke-AB5{stroke:#F7F8FE;}
		.d2-827229399 .background-color-N1{background-color:#0A0F25;}
		.d2-827229399 .background-color-N2{background-color:#676C7E;}
		.d2-827229399 .background-color-N3{background-color:#9499AB;}
		.d2-827229399 .background-color-N4{background-color:#CFD2DD;}
		.d2-827229399 .background-color-N5{background-color:#DEE1EB;}
		.d2-827229399 .background-color-N6{background-color:#EEF1F8;}
		.d2-827229399 .background-color-N7{background-color:#FFFFFF;}
		.d2-827229399 .background-color-B1{background-color:#0D32B2;}
		.d2-827229399 .background-color-B2{background-color:#0D32B2;}
		.d2-827229399 .background-color-B3{background-color:#E3E9FD;}
		.d2-827229399 .background-color-B4{background-color:#E3E9FD;}
		.d2-827229399 .background-color-B5{background-color:#EDF0FD;}
		.d2-827229399 .background-color-B6{background-color:#F7F8FE;}
		.d2-827229399 .background-color-AA2{background-color:#4A6FF3;}
		.d2-827229399 .background-color-AA4{background-color:#EDF0FD;}
		.d2-827229399 .background-color-AA5{background-color:#F7F8FE;}
		.d2-827229399 .background-color-AB4{background-color:#EDF0FD;}
		.d2-827229399 .background-color-AB5{background-color:#F7F8FE;}
		.d2-827229399 .color-N1{color:#0A0F25;}
		.d2-827229399 .color-N2{color:#676C7E;}
		.d2-827229399 .color-N3{color:#9499AB;}
		.d2-827229399 .color-N4{color:#CFD2DD;}
		.d2-827229399 .color-N5{color:#DEE1EB;}
		.d2-827229399 .color-N6{color:#EEF1F8;}
		.d2-827229399 .color-N7{color:#FFFFFF;}
		.d2-827229399 .color-B1{color:#0D32B2;}
		.d2-827229399 .color-B2{color:#0D32B2;}
		.d2-827229399 .color-B3{color:#E3E9FD;}
		.d2-827229399 .color-B4{color:#E3E9FD;}
		.d2-827229399 .color-B5{color:#EDF0FD;}
		.d2-827229399 .color-B6{color:#F7F8FE;}
		.d2-827229399 .color-AA2{color:#4A6FF3;}
		.d2-827229399 .color-AA4{color:#EDF0FD;}
		.d2-827229399 .color-AA5{color:#F7F8FE;}
		.d2-827229399 .color-AB4{color:#EDF0FD;}
		.d2-827229399 .color-AB5{color:#F7F8FE;}.appendix text.text{fill:#0A0F25}.md{--color-fg-default:#0A0F25;--color-fg-muted:#676C7E;--color-fg-subtle:#9499AB;--color-canvas-default:#FFFFFF;--color-canvas-subtle:#EEF1F8;--color-border-default:#0D32B2;--color-border-muted:#0D32B2;--color-neutral-muted:#EEF1F8;--color-accent-fg:#0D32B2;--color-accent-emphasis:#0D32B2;--color-attention-subtle:#676C7E;--color-danger-fg:red;}.sketch-overlay-B1{fill:url(#streaks-darker);mix-blend-mode:lighten}.sketch-overlay-B2{fill:url(#streaks-darker);mix-blend-mode:lighten}.sketch-overlay-B3{fill:url(#streaks-bright);mix-blend-mode:darken}.sketch-overlay-B4{fill:url(#streaks-bright);mix-blend-mode:darken}.sketch-overlay-B5{fill:url(#streaks-bright);mix-blend-mode:darken}.sketch-overlay-B6{fill:url(#streaks-bright);mix-blend-mode:darken}.sketch-overlay-AA2{fill:url(#streaks-dark);mix-blend-mode:overlay}.sketch-overlay-AA4{fill:url(#streaks-bright);mix-blend-mode:darken}.sketch-overlay-AA5{fill:url(#streaks-bright);mix-blend-mode:darken}.sketch-overlay-AB4{fill:url(#streaks-bright);mix-blend-mode:darken}.sketch-overlay-AB5{fill:url(#streaks-bright);mix-blend-mode:darken}.sketch-overlay-N1{fill:url(#streaks-darker);mix-blend-mode:lighten}.sketch-overlay-N2{fill:url(#streaks-dark);mix-blend-mode:overlay}.sketch-overlay-N3{fill:url(#streaks-normal);mix-blend-mode:color-burn}.sketch-overlay-N4{fill:url(#streaks-normal);mix-blend-mode:color-burn}.sketch-overlay-N5{fill:url(#streaks-bright);mix-blend-mode:darken}.sketch-overlay-N6{fill:url(#streaks-bright);mix-blend-mode:darken}.sketch-overlay-N7{fill:url(#streaks-bright);mix-blend-mode:darken}.light-code{display: block}.dark-code{display: none}]]></style><g id="module_policy_aws_iam_policy_policy"><g class="shape" ><rect x="0.000000" y="197.000000" width="160.000000" height="66.000000" class=" stroke-B1 fill-B6" style="stroke-width:2;" /></g><text x="80.000000" y="235.500000" class="text-bold fill-N1" style="text-anchor:middle;font-size:16px">aws_iam_policy</text></g><g id="module_roleA_aws_iam_role_this"><g class="shape" ><image href="https://raw.githubusercontent.com/tf2d2/icons/main/aws/resource/Security-Identity-Compliance/AWS-Identity-Access-Management_Role.svg" x="220.000000" y="166.000000" width="128.000000" height="128.000000" class=" stroke-B1 fill-N7" style="stroke-width:2;" /></g><text x="284.000000" y="315.000000" class="text-bold fill-N1" style="text-anchor:middle;font-size:16px">aws_iam_role</text></g><g id="module_roleA_aws_iam_role_policy_attachment_custom"><g class="shape" ><rect x="39.000000" y="0.000000" width="286.000000" height="66.000000" class=" stroke-B1 fill-B6" style="stroke-width:2;" /></g><text x="182.000000" y="38.500000" class="text-bold fill-N1" style="text-anchor:middle;font-size:16px">aws_iam_role_policy_attachment</text></g><g id="(module_roleA_aws_iam_role_policy_attachment_custom -&gt; module_policy_aws_iam_policy_policy)[0]"><marker id="mk-3488378134" markerWidth="10.000000" markerHeight="12.000000" refX="7.000000" refY="6.000000" viewBox="0.000000 0.000000 10.000000 12.000000" orient="auto" markerUnits="userSpaceOnUse"> <polygon points="0.000000,0.000000 10.000000,6.000000 0.000000,12.000000" class="connection fill-B1" stroke-width="2" /> </marker><path d="M 139.453203 67.267840 C 92.198997 106.000000 80.000000 132.199997 80.000000 193.000000" fill="none" class="connection stroke-B1" style="stroke-width:2;" marker-end="url(#mk-3488378134)" mask="url(#d2-827229399)" /></g><g id="(module_roleA_aws_iam_role_policy_attachment_custom -&gt; module_roleA_aws_iam_role_this)[0]"><path d="M 224.546784 67.267856 C 271.799988 106.000000 284.000000 126.000000 284.000000 162.000000" fill="none" class="connection stroke-B1" style="stroke-width:2;" marker-end="url(#mk-3488378134)" mask="url(#d2-827229399)" /></g><mask id="d2-827229399" maskUnits="userSpaceOnUse" x="-101" y="-101" width="550" height="521">
<rect x="-101" y="-101" width="550" height="521" fill="white"></rect>
<rect x="22.500000" y="219.500000" width="115" height="21" fill="rgba(0,0,0,0.75)"></rect>
<rect x="234.500000" y="299.000000" width="99" height="21" fill="rgba(0,0,0,0.75)"></rect>
<rect x="61.500000" y="22.500000" width="241" height="21" fill="rgba(0,0,0,0.75)"></rect>
</mask></svg></svg>
